traceutil/roundtrip.go

```diff
@@ -0,0 +1,191 @@
+package traceutil
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"net/http/httptrace"
+	"sync"
+	"time"
+)
+
+// RoundTripResult is a set of timestamps set at various stages of an outgoing
+// HTTP request intended to be passed to a FinishFunc passed in to RoundTrip.
+type RoundTripResult struct {
+	Started              time.Time
+	WroteHeaders         time.Time
+	WroteRequest         time.Time
+	GotFirstResponseByte time.Time
+	ReadHeaders          time.Time
+	ReadResponse         time.Time
+
+	BytesWritten int // TODO: implemenent
+	BytesRead    int
+
+	// The request traced. It is not safe to modify nor read from its body
+	// or headers.
+	Request *http.Request
+
+	// The response traced. It is not safe to modify nor read from its body
+	// of headers.  It is nil if no response was recieved.
+	Response *http.Response
+
+	Err error
+}
+
+func propagateHTTP(req *http.Request) {
+	// TODO: implement
+}
+
+type traceRoundTripper struct {
+	inner    http.RoundTripper
+	finished FinishFunc
+}
+
+// RoundTrip returns a http.RoundTripper that wraps inner for purposes of
+// recording timestamps to a RoundTripResult which is passed to finished on
+// completion or error of the returned http.RoundTripper.
+func RoundTrip(inner http.RoundTripper, finished FinishFunc) http.RoundTripper {
+	return traceRoundTripper{inner: inner, finished: finished}
+}
+
+type FinishFunc func(result *RoundTripResult)
+
+// FinishBasic records a single span for the request representing the from
+// first request byte written to last response byte read.
+func FinishBasic() FinishFunc {
```

Comment:
s/FinishBasic/Finish

Let's keep the naming simple, any additional finishers can extend the name.